### PR TITLE
[JENKINS-18114] Fixing CLI crumbs

### PR DESCRIPTION
**Pull request by jglick**
Amends #2315. The CLI client was already asking for a crumb. But Jenkins was not properly serving `X-Jenkins` alongside it, so it did not pass it back; but then Jenkins forgave it anyway! Simpler and clearer to just serve the expected header.

### Proposed changelog entries

None _should_ be needed.

@reviewbybees

### CODE REVIEW

**Comment by daniel-beck** [comment]
Test failures seem related.

**Comment by jglick** [comment]
`CLIActionTest.serveCliActionToAnonymousUserWithoutPermissions` was meaningless. But `CLIActionTest.authentication` was real: the proposed change actually breaks the (admittedly dubious) use case of running CLI commands as anonymous when anonymous has no Overall/Read.

Perhaps we need to return to the crumb exclusion, and just drop the client code to send the crumb, which adds to the round-trip time; is not needed when there is an exclusion; and in `master` does not even work anyway.

**Comment by jglick** [comment]
(Note that `FullDuplexHttpStream.tryToResolveRedirects` does not work for an anonymous user without read access either, but this is just a convenience in case you have a noncanonical Jenkins URL.)

**Comment by daniel-beck** [comment]
> admittedly dubious

This may be a symptom of a larger possible problem: Unless we want all UnprotectedRootActions to be CrumbExclusions, there probably should be a way for unauthorized users to get a crumb. Searching I now only found JENKINS-31515, but it's still something we should think about.

**Comment by jglick** [comment]
I do not think we can (safely) create a crumb for anonymous users. There is [JENKINS-22474](https://issues.jenkins-ci.org/browse/JENKINS-22474) which could allow us to delete most `CrumbExclusion`s for `RootAction`s. Arguably `UnprotectedRootAction`s should also be excluded automatically.

File: test/src/test/java/hudson/cli/CLIActionTest.java [hunk 1]
```diff
@@ -122,16 +117,6 @@ public void serveCliActionToAnonymousUserWithoutPermissions() throws Exception {
         JenkinsRule.WebClient wc = j.createWebClient();
         // The behavior changed due to SECURITY-192. index page is no longer accessible to anonymous
         wc.assertFails("cli", HttpURLConnection.HTTP_FORBIDDEN);
-        // so we check the access by emulating the CLI connection post request
```
Comment (jglick) [inline test/src/test/java/hudson/cli/CLIActionTest.java]: Actually would pass again given the restored crumb exclusion, but is anyway redundant given more comprehensive, and realistic, tests later added to the CLI system.

File: test/src/test/java/jenkins/util/FullDuplexHttpServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+public class FullDuplexHttpServiceTest {
+
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    @Rule
+    public LoggerRule logging = new LoggerRule().record(FullDuplexHttpService.class, Level.FINE).record(FullDuplexHttpStream.class, Level.FINE);
+
+    @Test
+    public void smokes() throws Exception {
```
Comment (jglick) [inline test/src/test/java/jenkins/util/FullDuplexHttpServiceTest.java]: Just for fun!

File: test/src/test/java/jenkins/util/FullDuplexHttpServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        logging.record("org.eclipse.jetty", Level.ALL);
```
Comment (jglick) [inline test/src/test/java/jenkins/util/FullDuplexHttpServiceTest.java]: I have been investigating failures of HTTP Duplex to work in test cases using Jetty 9.4.x. Symptoms are typically a failure to connect the upload side, as in [JENKINS-43666](https://issues.jenkins-ci.org/browse/JENKINS-43666).

Comment (jglick) [reply test/src/test/java/jenkins/util/FullDuplexHttpServiceTest.java]: (Starting from https://github.com/jenkinsci/jenkins/pull/2959 which picks up https://github.com/jenkinsci/jenkins-test-harness/pull/63. Seems unrelated to https://github.com/jenkinsci/jenkins/pull/3011.)

Comment (daniel-beck) [reply test/src/test/java/jenkins/util/FullDuplexHttpServiceTest.java]: How verbose is this? Intended to stay on ALL?

**Review by daniel-beck — APPROVED** [review]

**Comment by jglick** [comment]
@reviewbybees done

**Review by oleg-nenashev — APPROVED** [review]
:bee: IIUC

File: test/src/test/java/jenkins/util/FullDuplexHttpServiceTest.java (hunk 3 above)
Comment (jglick) [reply test/src/test/java/jenkins/util/FullDuplexHttpServiceTest.java]: In practice Jetty logs everything at `FINE`.

**Pull request merged**